File: src/registry.ts
import { readFileSync } from "node:fs";
import { z } from "zod";
import { BANK_IDS, type WriterRegistry, type WriterRule } from "./types.js";

export const DEFAULT_REGISTRY: WriterRegistry = {
  writers: {
    main: {
      role: "orchestrator",
      source: "application",
      write_bank: "main",
      read_banks: ["main", "core", "ops", "dev", "creative", "personal"],
    },
    ops: {
      role: "ops",
      source: "application",
      write_bank: "ops",
      read_banks: ["ops", "core"],
    },
    dev: {
      role: "dev",
      source: "application",
      write_bank: "dev",
      read_banks: ["dev", "core"],
    },
    creative: {
      role: "creative",
      source: "application",
      write_bank: "creative",
      read_banks: ["creative", "core"],
    },
    personal: {
      role: "personal",
      source: "application",
      write_bank: "personal",
      read_banks: ["personal", "core"],
    },
    research: {
      role: "research",
      source: "application",
      write_bank: "research",
      read_banks: ["research", "core"],
    },
  },
  defaults: {
    unknown_writer_action: "review_queue",
    suspicious_content_action: "review_queue",
  },
};

const BANK_ID_SET = new Set<string>(BANK_IDS);

const registrySchema = z
  .object({
    writers: z.record(z.string(), z.unknown()),
    defaults: z
      .object({
        unknown_writer_action: z.literal("review_queue"),
        suspicious_content_action: z.literal("review_queue"),
      })
      .passthrough(),
  })
  .passthrough();

const writerRuleSchema = z
  .object({
    role: z.string().refine((value) => value.trim().length > 0),
    source: z.string().refine((value) => value.trim().length > 0),
    write_bank: z.string().min(1),
    read_banks: z.array(z.string()),
  })
  .passthrough();

export function loadRegistry(path?: string): WriterRegistry {
  if (!path) return DEFAULT_REGISTRY;
  const parsed = JSON.parse(readFileSync(path, "utf8")) as unknown;
  validateRegistry(parsed as WriterRegistry);
  return parsed as WriterRegistry;
}

export function getWriter(
  registry: WriterRegistry,
  writerId: string,
): WriterRule | undefined {
  return registry.writers[writerId];
}

export function validateRegistry(registry: WriterRegistry): void {
  const parsed = registrySchema.safeParse(registry);
  if (!parsed.success) {
    throw new Error(
      registryValidationMessage(registry, parsed.error.issues[0]),
    );
  }

  for (const [writerId, value] of Object.entries(parsed.data.writers)) {
    if (!writerId.trim()) throw new Error("writer id cannot be empty");

    const writer = writerRuleSchema.safeParse(value);
    if (!writer.success) {
      throw new Error(
        writerValidationMessage(writerId, value, writer.error.issues[0]),
      );
    }

    const rule = writer.data;
    if (rule.write_bank === "quarantine") {
      throw new Error(`writer ${writerId} cannot write quarantine`);
    }
    if (!BANK_ID_SET.has(rule.write_bank)) {
      throw new Error(`writer ${writerId} has invalid write_bank`);
    }
    if (rule.read_banks.includes("quarantine")) {
      throw new Error(`writer ${writerId} cannot read quarantine`);
    }
    for (const bank of rule.read_banks) {
      if (!BANK_ID_SET.has(bank)) {
        throw new Error(`writer ${writerId} has invalid read_bank`);
      }
    }
    if (writerId === "main" && rule.read_banks.includes("research")) {
      throw new Error("main writer cannot read research");
    }
  }
}

function registryValidationMessage(
  value: unknown,
  issue: z.core.$ZodIssue | undefined,
): string {
  if (!isObject(value)) return "registry must be an object";
  if (issue?.path[0] === "writers") return "registry.writers must be an object";
  if (issue?.path[0] === "defaults") {
    if (issue.path[1] === "unknown_writer_action") {
      return "registry.defaults.unknown_writer_action must be review_queue";
    }
    if (issue.path[1] === "suspicious_content_action") {
      return "registry.defaults.suspicious_content_action must be review_queue";
    }
    return "registry.defaults must be an object";
  }
  return "registry must be an object";
}

function writerValidationMessage(
  writerId: string,
  value: unknown,
  issue: z.core.$ZodIssue | undefined,
): string {
  if (!isObject(value)) return `writer ${writerId} must be an object`;
  switch (issue?.path[0]) {
    case "role":
      return `writer ${writerId} missing role`;
    case "source":
      return `writer ${writerId} missing source`;
    case "write_bank":
      return `writer ${writerId} missing write_bank`;
    case "read_banks":
      return issue.path.length > 1
        ? `writer ${writerId} has invalid read_bank`
        : `writer ${writerId} missing read_banks`;
    default:
      return `writer ${writerId} must be an object`;
  }
}

function isObject(value: unknown): value is Record<string, unknown> {
  return value !== null && typeof value === "object" && !Array.isArray(value);
}

File: memory_router/registry.py
from __future__ import annotations

import json
from pathlib import Path
from typing import Any

from pydantic import ValidationError

from .models import BANK_IDS, WriterRegistry

DEFAULT_REGISTRY = WriterRegistry.model_validate(
    {
        "writers": {
            "main": {
                "role": "default",
                "source": "application",
                "write_bank": "main",
                "read_banks": ["main"],
            }
        },
        "defaults": {
            "unknown_writer_action": "review_queue",
            "suspicious_content_action": "review_queue",
        },
    }
)


def load_registry(path: str | None = None) -> WriterRegistry:
    if not path:
        return DEFAULT_REGISTRY.model_copy(deep=True)
    value = json.loads(Path(path).read_text(encoding="utf-8"))
    return validate_registry(value)


def validate_registry(value: Any) -> WriterRegistry:
    if not isinstance(value, dict):
        raise ValueError("registry must be an object")
    if not isinstance(value.get("writers"), dict):
        raise ValueError("registry.writers must be an object")
    defaults = value.get("defaults")
    if not isinstance(defaults, dict):
        raise ValueError("registry.defaults must be an object")
    if defaults.get("unknown_writer_action") != "review_queue":
        raise ValueError("registry.defaults.unknown_writer_action must be review_queue")
    if defaults.get("suspicious_content_action") != "review_queue":
        raise ValueError("registry.defaults.suspicious_content_action must be review_queue")

    for writer_id, raw in value["writers"].items():
        if not isinstance(writer_id, str) or not writer_id.strip():
            raise ValueError("writer id cannot be empty")
        if not isinstance(raw, dict):
            raise ValueError(f"writer {writer_id} must be an object")
        for field in ("role", "source", "write_bank", "read_banks"):
            if field not in raw:
                raise ValueError(f"writer {writer_id} missing {field}")
        if not isinstance(raw["role"], str) or not raw["role"].strip():
            raise ValueError(f"writer {writer_id} missing role")
        if not isinstance(raw["source"], str) or not raw["source"].strip():
            raise ValueError(f"writer {writer_id} missing source")
        if raw["write_bank"] == "quarantine":
            raise ValueError(f"writer {writer_id} cannot write quarantine")
        if raw["write_bank"] not in BANK_IDS:
            raise ValueError(f"writer {writer_id} has invalid write_bank")
        if not isinstance(raw["read_banks"], list):
            raise ValueError(f"writer {writer_id} missing read_banks")
        if "quarantine" in raw["read_banks"]:
            raise ValueError(f"writer {writer_id} cannot read quarantine")
        if any(bank not in BANK_IDS for bank in raw["read_banks"]):
            raise ValueError(f"writer {writer_id} has invalid read_bank")
        if writer_id == "main" and "research" in raw["read_banks"]:
            raise ValueError("main writer cannot read research")
    try:
        return WriterRegistry.model_validate(value)
    except ValidationError as exc:
        raise ValueError("registry must be an object") from exc
